src/logger.c

```diff
@@ -136,24 +137,37 @@ int log_lock(int fd, int cmd, int type, const char *name,
 	        : cmd == F_UNLCK ? "F_UNLCK" : "");
 #endif
 
+	uint64_t lock_start = 0;
+	int warned = 0;
+	int retry = 0;
+
 	lock.l_type = type;
 	while (fcntl(fd, cmd, &lock)) {
-		if (errno == EAGAIN) {
-			static int warned;
+		if (errno == EAGAIN || errno == ENOLCK) {
 			struct timeval t;
+			unsigned int warn_bit = (errno == EAGAIN ? 1 : 2);
 
 			if (cmd == F_SETLK)
 				return -1;
 
-			if (!warned++) {
-				log_event("Got EAGAIN despite F_SETLKW (only logged once per node) %s:%d %s", source_file, line, function);
-				fprintf(stderr, "Node %d: File locking apparently exhausted, check ulimits and any NFS server limits. This is recoverable but will harm performance (muting further of these messages from same node)\n", NODE);
+			if (!(warned & warn_bit)) {
+				if (!lock_start)
+					lock_start = john_get_nano();
+
+				log_event("- Got %s despite F_SETLKW on %s, %s() %s:%d", errno == EAGAIN ? "EAGAIN" : "ENOLCK", name, function, source_file, line);
+				if (options.node_count)
+					fprintf(stderr, "%u: ", NODE);
+				fprintf(stderr, "File locking apparently exhausted (%s), retrying. Check ulimits and any NFS server limits\n", strerror(errno));
 				srand(NODE);
```

Comment:
I suggest dropping the `srand(NODE);` line from here. I think it actually hurts. Previously, it was only reached once because the `warned` variable was static - but now that you make the warning bits per-call, allowing for repeated log messages from subsequent calls, you'd also end up resetting the PRNG to the same state multiple times.

Reply:
Well, maybe limit the `srand(NODE);` to be done just once per process like it was before?